login/login.go

```diff
@@ -0,0 +1,390 @@
+	}
+	log.Println("info: ", info)
+
+	return DeviceLogin{BaseURL: info.OidcURI, ClientID: info.ClientID, PollingInterval: 2, S3Target: info.InboxURI}, nil
+}
+
+// Login() does a full login by fetching the remote configuration, starting the
+// login procedure, and then waiting for the user to complete login.
+func (login *DeviceLogin) Login() error {
+
+	var err error
+	login.wellKnown, err = login.getWellKnown()
+	if err != nil {
+		return fmt.Errorf("failed to fetch .well-known configuration: %v", err)
+	}
+
+	login.deviceLogin, err = login.startDeviceLogin()
+	if err != nil {
+		return fmt.Errorf("failed to start device login: %v", err)
+	}
+	expires := time.Duration(login.deviceLogin.ExpiresIn * int(time.Second))
+	log.Infof("Login started (expires in %v minutes)", expires.Minutes())
+	log.Infof("Go to %v to finish logging in.", login.deviceLogin.VerificationURL)
+
+	loginResult, err := login.waitForLogin()
+	if err != nil {
+		return err
+	}
+	login.LoginResult = loginResult
+
+	login.UserInfo, err = login.getUserInfo()
+	if err != nil {
+		return err
+	}
+
+	err = login.Session()
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// S3Config() returns an
+func (login *DeviceLogin) GetS3Config() (*S3Config, error) {
+	if login.LoginResult.AccessToken == "" {
+
+		return nil, errors.New("no login token available for config")
+	}
+
+	return &S3Config{
+		AccessKey:            login.UserInfo.Sub,
+		SecretKey:            login.UserInfo.Sub,
+		AccessToken:          login.LoginResult.AccessToken,
+		HostBucket:           login.S3Target,
+		HostBase:             login.S3Target,
+		MultipartChunkSizeMb: 512,
+		GuessMimeType:        false,
+		Encoding:             "UTF-8",
+		CheckSslCertificate:  false,
+		CheckSslHostname:     false,
+		UseHTTPS:             true,
+		SocketTimeout:        30,
+		HumanReadableSizes:   true,
+	}, nil
+}
+
+func (login *DeviceLogin) getUserInfo() (*UserInfo, error) {
+
+	if login.LoginResult.AccessToken == "" {
+		return nil, errors.New("login token required to fetch userinfo")
+	}
+
+	req, err := http.NewRequest("GET", login.BaseURL+"/userinfo", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %v", login.LoginResult.AccessToken))
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
+		err = fmt.Errorf("status code: %v", resp.StatusCode)
+
+		return nil, fmt.Errorf("request failed: %v", err)
+	}
+
+	var userinfo *UserInfo
+	err = json.Unmarshal(body, &userinfo)
+
+	return userinfo, err
+}
+
+// getWellKnown() makes a GET request to the `.well-known/openid-configuration`
+// endpoint of BaseURL and returns the result as `OIDCWellKnown`.
+func (login *DeviceLogin) getWellKnown() (*OIDCWellKnown, error) {
+	login.BaseURL = "https://login.elixir-czech.org/oidc"
```

Comment:
```suggestion
```
This should not be hardcoded

Reply:
right! testing leftovers. will remove it